### PR TITLE
Test construction of ocsp requests with "-cert_chain"

### PR DESCRIPTION
**Pull request by jamuir**
In #22192, functionality was added to allow `s_server` to construct ocsp requests using the option `-cert_chain`.  See the conversation there for more details.

Add a test for that.

You can exercise the test like so:

```
make V=1 TESTS='test_ocsp_cert_chain' test
```

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by jamuir** [comment]
Note that the changes to `apps/s_server.c` were done by olszomal and come from #22192.

Any comments on  `apps/s_server.c` should be made in the other PR.

**Comment by jamuir** [comment]
There are some CI failures.

I used ec certs, so I will have to skip this test for "no-ec".

**Comment by t8m** [comment]
CI still failing

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (t8m) [inline test/recipes/82-test_ocsp_cert_chain.t:45]: These fixed port numbers are problematic when the build is simultaneously running multiple times on the same machine. Could you please make the port number here and below a random one from some range (i.e., 19200-19299 should be good enough).

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (t8m) [inline test/recipes/82-test_ocsp_cert_chain.t:62]: Same here.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (t8m) [reply test/recipes/82-test_ocsp_cert_chain.t:45]: Although it will be complicating the test quite significantly as the certificate will have to be generated on-fly.

Comment (t8m) [reply test/recipes/82-test_ocsp_cert_chain.t:45]: Alternatively we could check here for the failure of starting the ocsp responder on the fixed port and just skip the test in such case. That could be easier to do.

The s_server port could be still randomized.

Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:45]: thanks for your comments, Tomas.

I will randomize the s_server port.

I agree that randomizing the ocsp port would be cumbersome.  I did include a script to regenerate the various keys and certs, but randomizing the ocsp port would seem to require updating `ca.cnf` (unless there is a way to pass that info via an option).

Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:53]: > check here for the failure of starting the ocsp responder on the fixed port and just skip the test in such case.

The danger with skipping is that the port the ocsp responder wants might become unavailable permanently (e.g. maybe it is claimed by another test).  The result would be that the test stops being run and we do not notice because there is no failure message.

**Comment by jamuir** [comment]
CI fails on the free-bsd node.

`make test` seems to be hung.

maybe the while-loops used in the new test need some limits applied to them.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:53]: We discussed the busy-port problem at the stand-up today.

Matt C made the suggestion to introduce a change to `s_server` that allows an environment variable to override the OCSP url from the cert.  We could use that to direct the OCSP query to the correct port.

Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t]: after looking at the s_server code, I saw there already exists an option `-status_url` that can be used to specify the url for the ocsp responder when s_server is run.  That url is used when 
the `authorityInfoAccess` extension is not present in the server cert.

I've updated the test to use `-status_url` and now both ports can be randomized.

**Comment by jamuir** [comment]
This is ready for review now (both ocsp-responder and tls-server use random ports).

Note that I force-pushed to the branch (hopefully, this does not make it more difficult to review).

**Comment by jamuir** [comment]
The three CI failures were due to ipv6 vs ipv4 (i.e. the string `::` vs `0.0.0.0`)

**Review by t8m — APPROVED** [review]

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [inline test/recipes/82-test_ocsp_cert_chain.t:93]: Isn't this going to cause the test to fail in the event that the port is already in use?

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [inline test/recipes/82-test_ocsp_cert_chain.t:99]: Using a server port of ":0" will cause the OS to allocate a random free port for you, avoiding the risk of randomly selecting an already in use port. Not sure if it is possible to use the same trick with the ocsp server somehow.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [inline test/recipes/82-test_ocsp_cert_chain.t:100]: This is going to cause problems. Some machines that we build on do not have IPv4 available!!

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [inline test/recipes/82-test_ocsp_cert_chain.t:122]: ??? I'm not sure what 0.0.0.0 means from a client perspective...in any case it will cause problems on an IPv6 only machine.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:93]: yes.  That was my intention.  I want the test to fail if the port is already in use.

skipping or passing the test if the port is busy did not seem like a good idea to me.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:100]: So on those machines, the loopback interface does not recognize 127.0.0.1?  Is that correct?

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:99]: > avoiding the risk of randomly selecting an already in use port.

If the risk of selecting a port that is already in use is very low, then we do not have to worry too much about it. 

The main thing I was concerned about was running multiple instances of the openssl tests on the same machine.

I confirmed that using `openssl s_server -accept :0 ...` results in a random port being used on my ubuntu machine.  Dropping the colon works, too:  `openssl s_server -accept 0 ...`

For ocsp, using `:0` results in a parse error, but `0` works: `openssl ocsp -port 0 ...`

This behaviour seems to be OS dependent (I cannot find any mention of it in the openssl docs).

I am hesitant to use port 0 if it is OS dependent.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:122]: on linux, `0.0.0.0` can apparently be used like `localhost`:

https://en.wikipedia.org/wiki/0.0.0.0

It is probably better to use `localhost` instead.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [reply test/recipes/82-test_ocsp_cert_chain.t:93]: Hmmm, I think that is problematic. It may well cause spurious, difficult to investigate failures for our users - for issues which are ultimately environmental and not problems in OpenSSL itself. I would prefer to use an approach where we always pick a port that is known to be free - or failing that, that we skip a test in the event of such a failure.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [reply test/recipes/82-test_ocsp_cert_chain.t:99]: > If the risk of selecting a port that is already in use is very low, then we do not have to worry too much about it.

I don't really agree with this statement. Our tests get repeatedly run in many different environments by many different users. If enough users run the tests in enough different environments then even low probability events can occur and can cause difficult to investigate and diagnose problems.

> I am hesitant to use port 0 if it is OS dependent.

We are already using this trick in other tests:

https://github.com/openssl/openssl/blob/3348713ad390372ba5a0a0f98b46b2f637475e47/util/perl/TLSProxy/Proxy.pm#L222-L232

I don't recall any reports of issues relating to this.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (mattcaswell) [reply test/recipes/82-test_ocsp_cert_chain.t:100]: Right

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t:99]: > If enough users run the tests in enough different environments then even low probability events can occur

It depends on the magnitude of that low probability event, which, admittedly, I have been vague about :-)

e.g. probable-primes are used to generate public-key params.

> We are already using this trick in other tests

Thanks for pointing to that test.  Since no one has complained about port 0, I will push an update that uses it.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t]: the behaviour has been changed.  please see the latest commit.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t]: I am now using port 0.  Please see the latest changes.

**Review by t8m — APPROVED** [review]

**Comment by jamuir** [comment]
I hope Matt will approve this latest version :-)

Remember that this PR is dependent on #22995, so it should only be merged after #22995 goes in.

Further, this PR is currently based on commit 855bdea3 (from #22995), but that commit is out of date (the new commit is c7201394).

**Comment by jamuir** [comment]
I see that #22995 has been merged.

@mattcaswell :  do you have a few minutes to look at the changes made in response to your review?

After this is approved, I can rebase the branch to simplify the merge (remember that commit 855bdea is out of date).

**Review by mattcaswell — APPROVED** [review]
> @mattcaswell : do you have a few minutes to look at the changes made in response to your review?

Sorry - I somehow missed this ping.

Anyway, approved, subject to #22192 also getting approved.

**Comment by t8m** [comment]
This is on hold for a little bit different reason. But I've needed to apply some label so it does not get accidentally merged.

File: test/recipes/82-test_ocsp_cert_chain.t
Comment (Tarnum-tst) [inline test/recipes/82-test_ocsp_cert_chain.t:29]: Please add msys here. This test hangs on MSYS over MSWin32 platform too.

Comment (jamuir) [reply test/recipes/82-test_ocsp_cert_chain.t]: thanks for catching this.

I've just pushed an update.

**Review by shahsb — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by FdaSilvaYY** [comment]
Ready to merge ?

**Comment by t8m** [comment]
This is waiting on #22192 to be merged

**Comment by t8m** [comment]
@jamuir Could you please rebase this now that #22192 was merged?

**Comment by jamuir** [comment]
> @jamuir Could you please rebase this now that #22192 was merged?

done!

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]
Reconfirm

**Comment by t8m** [comment]
Merged to the master branch.

**Comment by t8m** [comment]
@jamuir It seems this is breaking on the AFL CI job: https://github.com/openssl/openssl/actions/runs/8254459118/job/22579062368 Any idea?

**Comment by jamuir** [comment]
No, off-hand I don't see why it would be failing with AFL.

Here is the important part of the log:

```
s_server ready, listening on port 35063
Warning: AFL++ tools might need to set AFL_MAP_SIZE to 120985 to be able to run this instrumented program if this crashes!
not ok 3 - check s_server sent ocsp response
# ------------------------------------------------------------------------------
#   Failed test 'check s_server sent ocsp response'
#   at test/recipes/82-test_ocsp_cert_chain.t line 137.
# Looks like you failed 1 test of 3.82-test_ocsp_cert_chain.t ............... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/3 subtests 
```

I don't know if the `Warning: AFL...` message is significant.

That part of the test checks stdout for a certain string (like the two checks before it inside `82-test_ocsp_cert_chain.t`)

Is it possible that the fuzzing instrumentation somehow causes another string to be output before the one we are looking for?

If someone could run the test manually using AFL, then we could check that.

**Comment by jamuir** [comment]
> I don't know if the `Warning: AFL...` message is significant.

On second thought, it is likely the warning message itself that is breaking the test.

i.e. we were looking for the string `cert_status: ocsp response sent:`, but we instead got the string `Warning: AFL++ tools might need...` so we fail.

**Comment by t8m** [comment]
I've created https://github.com/openssl/openssl/pull/23843 although it is not really a crash so the commit message is misleading. I'll fix that when merging.

**Pull request closed without merging**
